src/System.Management.Automation/engine/pipeline.cs

```diff
@@ -1182,27 +1230,27 @@ private void Inject(object input, bool enumerate)
         /// </returns>
         private Array RetrieveResults()
         {
+            if (_commands is null)
+            {
+                // This could happen to an expression redirection pipeline (e.g. 1 > a.txt).
+                // An exception may be thrown from the call to 'StartStepping' or 'Step' on the pipeline,
+                // which causes the pipeline commands to be disposed.
+                return MshCommandRuntime.StaticEmptyArray;
+            }
+
             // If the error queue has been linked, it's up to the link to
             // deal with the output. Don't do anything here...
             if (!_linkedErrorOutput)
             {
-                // Retrieve any accumulated error objects from each of the pipes
-                // and add them to the error results hash table.
                 for (int i = 0; i < _commands.Count; i++)
                 {
                     CommandProcessorBase commandProcessor = _commands[i];
-                    if (commandProcessor == null
-                        || commandProcessor.CommandRuntime == null)
-                    {
-                        // "null command or request or ErrorOutputPipe " + i
-                        throw PSTraceSource.NewInvalidOperationException();
-                    }
+                    ValidateCommandProcessorNotNull(commandProcessor, errorMessage: null);
```

Comment:
I wonder you do not change to `foreach`.

Reply:
Updated.